### PR TITLE
Add free join-semilattices, free meet-semilattices and free lattices

### PR DESCRIPTION
**Pull request by ocharles**
How does this look?

### CODE REVIEW

File: Algebra/Lattice/Free.hs [hunk 1]
```diff
@@ -48,7 +52,9 @@ retractFreeJoinSemiLattice a = lowerFreeJoinSemiLattice a id
 
 instance Functor FreeJoinSemiLattice where
   fmap f (FreeJoinSemiLattice g) = FreeJoinSemiLattice (\inj -> g (inj . f))
+#if MIN_VERSION_base(4,2,0)
```
Comment (phadej) [inline Algebra/Lattice/Free.hs]: lattices have `base >= 4.5` so CPP isn't required, import should be enough

Comment (ocharles) [inline Algebra/Lattice/Free.hs]: Hmm, I had to do this as the Travis build failed otherwise

Comment (phadej) [inline Algebra/Lattice/Free.hs]: Ok, I'll check what's wrong. Will merge soon-ish.

Comment (phadej) [inline Algebra/Lattice/Free.hs]: which job failed?

Comment (ocharles) [inline Algebra/Lattice/Free.hs]: See https://travis-ci.org/phadej/lattices/builds/220532064

**Pull request closed without merging**